lib/ansible/modules/cloud/amazon/lambda_policy.py

```diff
@@ -0,0 +1,529 @@
+
+def get_policy_statement(module, aws):
+    """
+    Checks that policy exists and if so, that statement ID is present or absent.
+
+    :param module:
+    :param aws:
+    :return:
+    """
+
+    client = aws.client('lambda')
+    policy = dict()
+    policy_statement = dict()
+    sid = module.params['statement_id']
+
+    # set API parameters
+    api_params = set_api_params(module, ('function_name', ))
+    qualifier = get_qualifier(module)
+    if qualifier:
+        api_params.update(Qualifier=qualifier)
+
+    # check if function policy exists
+    try:
+        # get_policy returns a JSON string so must convert to dict before reassigning to its key
+        policy_results = client.get_policy(**api_params)
+        policy = json.loads(policy_results.get('Policy', '{}'))
+
+    except Exception as e:
+        try:
+            rc = e.response['Error']['Code']
+            if rc != 'ResourceNotFoundException':
+                fail_json_aws(module, e, msg="retrieving function policy")
+        except Exception:
+            fail_json_aws(module, e, msg="retrieving function policy")
+
+    if 'Statement' in policy:
+        # Now that we have the policy, check if required permission statement is present and flatten to
+        # simple dictionary if found.
+        for statement in policy['Statement']:
+            if statement['Sid'] == sid:
+                policy_statement['action'] = statement['Action']
+                policy_statement['principal'] = statement['Principal']['Service']
+                try:
+                    policy_statement['source_arn'] = statement['Condition']['ArnLike']['AWS:SourceArn']
+                except KeyError:
+                    pass
+                try:
+                    policy_statement['source_account'] = statement['Condition']['StringEquals']['AWS:SourceAccount']
+                except KeyError:
+                    pass
+                try:
+                    policy_statement['event_source_token'] = statement['Condition']['StringEquals']['lambda:EventSourceToken']
+                except KeyError:
+                    pass
+                break
+
+    return policy_statement
+
+
+def add_policy_permission(module, aws):
+    """
+    Adds a permission statement to the policy.
+
+    :param module:
+    :param aws:
+    :return:
+    """
+
+    client = aws.client('lambda')
+    changed = False
+
+    # set API parameters
+    params = (
+        'function_name',
+        'statement_id',
+        'action',
+        'principal',
+        'source_arn',
+        'source_account',
+        'event_source_token')
+    api_params = set_api_params(module, params)
+    qualifier = get_qualifier(module)
+    if qualifier:
+        api_params.update(Qualifier=qualifier)
+
+    try:
+        if not module.check_mode:
+            client.add_permission(**api_params)
+        changed = True
+    except Exception as e:
```

Comment:
I know that in certain cases - e.g. when the network connection suddenly disappears half way through a connection - botocore/boto3 can throw non boto exceptions.  The current design of aws_module.fail_json_aws() is that it should be able to accept any reasonable exception so I think it's okay to catch Exception.  Otherwise, I think we have to work out a more or less complete list of the exceptions botocore can throw and provide them as a constant somewhere.  

For now, at least, I'll leave this as catching exception and let's have a discuss through e.g. on IRC